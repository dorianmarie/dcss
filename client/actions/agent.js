import {
  GET_AGENT_ERROR,
  GET_AGENT_SUCCESS,
  GET_AGENTS_ERROR,
  GET_AGENTS_SUCCESS,
  SET_AGENT_ERROR,
  SET_AGENT_SUCCESS
} from './types';

export let getAgent = id => async dispatch => {
  try {
    const res = await (await fetch(`/api/agents/${id}`)).json();

    if (res.error) {
      throw res;
    }
    const { agent } = res;

    dispatch({ type: GET_AGENT_SUCCESS, agent });
    return agent;
  } catch (error) {
    dispatch({ type: GET_AGENT_ERROR, error });
    return null;
  }
};

export let setAgent = (id, params) => async dispatch => {
  try {
    if (Object.values(params).length) {
      const body = JSON.stringify(params);
      const res = await (await fetch(`/api/agents/${id}`, {
        method: 'PUT',
        headers: {
          'Content-Type': 'application/json'
        },
        body
      })).json();

      if (res.error) {
        throw res;
      }
      const { agent } = res;
      dispatch({ type: SET_AGENT_SUCCESS, agent });
      return agent;
    }
    return null;
  } catch (error) {
    dispatch({ type: SET_AGENT_ERROR, error });
    return null;
  }
};

export let createAgent = params => async dispatch => {
  try {
    if (Object.values(params).length) {
      const body = JSON.stringify(params);
      const res = await (await fetch(`/api/agents`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json'
        },
        body
      })).json();

      if (res.error) {
        throw res;
      }
      const { agent } = res;
      dispatch({ type: GET_AGENT_SUCCESS, agent });
      return agent;
    }
    return null;
  } catch (error) {
    dispatch({ type: GET_AGENT_ERROR, error });
    return null;
  }
};

export let getAgents = filter => async dispatch => {
  try {
    const url =
      filter === 'is_active' ? '/api/agents/is_active' : '/api/agents';

    const res = await (await fetch(url)).json();

    if (res.error) {
      throw res;
    }

    const { agents = [] } = res;
    dispatch({ type: GET_AGENTS_SUCCESS, agents });
    return agents;
  } catch (error) {
    dispatch({ type: GET_AGENTS_ERROR, error });
    return null;
  }
};
