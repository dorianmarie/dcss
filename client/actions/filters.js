import { SET_FILTER_SCENARIOS_IN_USE } from './types';

export let setFilterScenariosInUse = scenariosInUse => dispatch => {
  dispatch({ type: SET_FILTER_SCENARIOS_IN_USE, scenariosInUse });
};
