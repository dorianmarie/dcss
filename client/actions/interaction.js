import {
  GET_INTERACTION_ERROR,
  GET_INTERACTION_SUCCESS,
  GET_INTERACTIONS_ERROR,
  GET_INTERACTIONS_SUCCESS,
  SET_INTERACTION_ERROR,
  SET_INTERACTION_SUCCESS
} from './types';

export let getInteraction = id => async dispatch => {
  try {
    const res = await (await fetch(`/api/interactions/${id}`)).json();

    if (res.error) {
      throw res;
    }
    const { agent } = res;

    dispatch({ type: GET_INTERACTION_SUCCESS, agent });
    return agent;
  } catch (error) {
    dispatch({ type: GET_INTERACTION_ERROR, error });
    return null;
  }
};

export let setInteraction = (id, params) => async dispatch => {
  try {
    if (Object.values(params).length) {
      const body = JSON.stringify(params);
      const res = await (await fetch(`/api/interactions/${id}`, {
        method: 'PUT',
        headers: {
          'Content-Type': 'application/json'
        },
        body
      })).json();

      if (res.error) {
        throw res;
      }
      const { agent } = res;
      dispatch({ type: SET_INTERACTION_SUCCESS, agent });
      return agent;
    }
    return null;
  } catch (error) {
    dispatch({ type: SET_INTERACTION_ERROR, error });
    return null;
  }
};

export let createInteraction = params => async dispatch => {
  try {
    if (Object.values(params).length) {
      const body = JSON.stringify(params);
      const res = await (await fetch(`/api/interactions`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json'
        },
        body
      })).json();

      if (res.error) {
        throw res;
      }
      const { agent } = res;
      dispatch({ type: GET_INTERACTION_SUCCESS, agent });
      return agent;
    }
    return null;
  } catch (error) {
    dispatch({ type: GET_INTERACTION_ERROR, error });
    return null;
  }
};

export let getInteractions = () => async dispatch => {
  try {
    const res = await (await fetch(`/api/interactions`)).json();

    if (res.error) {
      throw res;
    }
    const { interactions } = res;

    dispatch({ type: GET_INTERACTIONS_SUCCESS, interactions });
    return interactions;
  } catch (error) {
    dispatch({ type: GET_INTERACTIONS_ERROR, error });
    return null;
  }
};
