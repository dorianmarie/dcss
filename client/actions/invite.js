import {
  GET_INVITES_ERROR,
  GET_INVITES_SUCCESS,
  SET_INVITE_ERROR,
  SET_INVITE_SUCCESS
} from './types';

export let getInvites = () => async dispatch => {
  try {
    const res = await (await fetch(`/api/invites`)).json();

    if (res.error) {
      throw res;
    }
    const { invites } = res;
    dispatch({ type: GET_INVITES_SUCCESS, invites });
    return invites;
  } catch (error) {
    dispatch({ type: GET_INVITES_ERROR, error });
    return null;
  }
};

export let setInvite = (id, params) => async dispatch => {
  try {
    const { status } = params;
    const updates = {};

    if (status) {
      updates.status = status;
    }

    const body = JSON.stringify(updates);
    const res = await (await fetch(`/api/invites/${id}`, {
      method: 'PUT',
      headers: {
        'Content-Type': 'application/json'
      },
      body
    })).json();

    if (res.error) {
      throw res;
    }

    const { invites } = res;
    const invite = invites.find(invite => invite.id === id);

    dispatch({ type: SET_INVITE_SUCCESS, invite });
    return invite;
  } catch (error) {
    dispatch({ type: SET_INVITE_ERROR, error });
    return null;
  }
};
