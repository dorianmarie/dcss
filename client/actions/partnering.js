import { GET_PARTNERING_ERROR, GET_PARTNERING_SUCCESS } from './types';

export let getPartnering = () => async (dispatch, getState) => {
  const state = getState();

  if (state.partnering.length) {
    return state.partnering;
  }

  try {
    const res = await (await fetch('/api/partnering')).json();

    if (res.error) {
      throw res;
    }

    const { partnering } = res;

    dispatch({ type: GET_PARTNERING_SUCCESS, partnering });
    return partnering;
  } catch (error) {
    dispatch({ type: GET_PARTNERING_ERROR, error });
    return null;
  }
};
