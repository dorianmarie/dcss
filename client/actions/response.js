import Storage from '@utils/Storage';
import {
  GET_RESPONSE_SUCCESS,
  GET_RESPONSE_ERROR,
  GET_TRANSCRIPTION_OUTCOME_SUCCESS,
  GET_TRANSCRIPTION_OUTCOME_ERROR,
  SET_RESPONSES_SUCCESS,
  SET_RESPONSES_ERROR
} from './types';

export let getResponse = (id, responseId) => async dispatch => {
  try {
    const res = await (await fetch(
      `/api/runs/${id}/response/${responseId}`
    )).json();

    if (res.error) {
      throw res;
    }

    let { response } = res;

    if (!response) {
      response = { response: null };
    }
    dispatch({ type: GET_RESPONSE_SUCCESS, response });
    return response;
  } catch (error) {
    dispatch({ type: GET_RESPONSE_ERROR, error });
    return null;
  }
};

export let getTranscriptionOutcome = ({ id, responseId }) => async dispatch => {
  try {
    const res = await (await fetch(
      `/api/runs/${id}/response/${responseId}/transcript`
    )).json();

    if (res.error) {
      throw res;
    }
    const { outcome } = res;
    if (outcome) {
      dispatch({ type: GET_TRANSCRIPTION_OUTCOME_SUCCESS, outcome });
      return outcome;
    }
    return {
      response: null,
      transcript: null
    };
  } catch (error) {
    dispatch({ type: GET_TRANSCRIPTION_OUTCOME_ERROR, error });
    return null;
  }
};

export let setResponses = (id, submitted) => async dispatch => {
  const responses = [];
  const responsesById = {};
  try {
    for (let [responseId, body] of submitted) {
      const res = await (await fetch(`/api/runs/${id}/response/${responseId}`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json'
        },
        body: JSON.stringify(body)
      })).json();

      if (res.error) {
        throw res;
      }

      const { response } = res;

      responses.push(response);
      responsesById[responseId] = response;

      if (Storage.has(`run/${id}/${responseId}`)) {
        Storage.delete(`run/${id}/${responseId}`);
      }
    }
    dispatch({ type: SET_RESPONSES_SUCCESS, responses, responsesById });
    return responses;
  } catch (error) {
    dispatch({ type: SET_RESPONSES_ERROR, error });
    return null;
  }
};
