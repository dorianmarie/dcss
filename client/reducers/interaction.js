import {
  GET_INTERACTION_SUCCESS,
  SET_INTERACTION_SUCCESS,
  GET_INTERACTIONS_SUCCESS
} from '@actions/types';
import { interactionInitialState } from './initial-states';

export const interaction = (state = interactionInitialState, action) => {
  const { interaction, type } = action;
  switch (type) {
    case GET_INTERACTION_SUCCESS:
    case SET_INTERACTION_SUCCESS: {
      return {
        ...state,
        ...interaction
      };
    }
    default:
      return state;
  }
};

export const interactions = (state = [], action) => {
  const { interaction, interactions, type } = action;
  switch (type) {
    case GET_INTERACTION_SUCCESS:
    case SET_INTERACTION_SUCCESS: {
      const index = state.findIndex(({ id }) => id === interaction.id);
      if (index !== -1) {
        state[index] = {
          ...state[index],
          ...interaction
        };
      } else {
        state.push(interaction);
      }
      return [...state].sort((a, b) => a.id < b.id);
    }
    case GET_INTERACTIONS_SUCCESS: {
      const seen = {
        /*
        interaction.id: index
         */
      };
      return [...state, ...interactions]
        .reduce((accum, interaction) => {
          // If we've already seen this interaction, then it was already
          // in "state" and we should replace it with the newer version.
          const index = seen[interaction.id];
          if (index !== undefined) {
            accum[index] = interaction;
          } else {
            seen[interaction.id] = accum.length;
            accum.push(interaction);
          }
          return accum;
        }, [])
        .sort((a, b) => a.id < b.id);
    }
    default:
      return state;
  }
};

export const interactionsById = (state = {}, action) => {
  const { interaction, interactions, type } = action;
  switch (type) {
    case GET_INTERACTION_SUCCESS:
    case SET_INTERACTION_SUCCESS: {
      return {
        ...state,
        [interaction.id]: {
          ...interaction
        }
      };
    }
    case GET_INTERACTIONS_SUCCESS: {
      const interactionsById = interactions.reduce(
        (accum, interaction) => ({
          ...accum,
          [interaction.id]: interaction
        }),
        {}
      );
      return {
        ...state,
        ...interactionsById
      };
    }
    default:
      return state;
  }
};
