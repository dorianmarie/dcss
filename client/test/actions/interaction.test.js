import { v4 as uuid } from 'uuid';
import assert from 'assert';
import {
  createMockStore,
  createMockConnectedStore,
  fetchImplementation,
  makeById,
  state
} from '../bootstrap';

import * as actions from '../../actions/interaction';
import * as types from '../../actions/types';
import Storage from '../../util/Storage';
jest.mock('../../util/Storage');

const error = new Error('something unexpected happened on the server');
const original = JSON.parse(JSON.stringify(state));
let mockStore;
let store;

beforeAll(() => {
  (window || global).fetch = jest.fn();
  Storage.has = jest.fn();
  Storage.delete = jest.fn();
});

afterAll(() => {
  jest.restoreAllMocks();
  Storage.has.mockRestore();
  Storage.delete.mockRestore();
});

beforeEach(() => {
  mockStore = createMockStore({});
  store = createMockConnectedStore({});
  fetch.mockImplementation(() => {});
  Storage.has.mockImplementation(() => true);
  Storage.delete.mockImplementation(() => {});
});

afterEach(() => {
  jest.resetAllMocks();
  Storage.has.mockReset();
  Storage.delete.mockReset();
});

describe('GET_INTERACTION_SUCCESS', () => {
  describe('getInteraction', () => {
    let interaction = { ...state.interactions[0], name: 'Fake Interaction' };

    test('Receives a interaction', async () => {
      fetchImplementation(fetch, 200, { interaction });
      const returnValue = await store.dispatch(actions.getInteraction(1));
      expect(fetch.mock.calls.length).toBe(1);
      expect(fetch.mock.calls[0]).toMatchInlineSnapshot(`
        Array [
          "/api/interactions/1",
        ]
      `);
      expect(returnValue).toEqual(interaction);

      await mockStore.dispatch(actions.getInteraction(1));
      expect(mockStore.getActions()).toMatchSnapshot();
    });
  });

  describe('createInteraction', () => {
    let interaction = { ...state.interactions[0], name: 'Fake Interaction' };

    test('Receives a interaction', async () => {
      fetchImplementation(fetch, 200, { interaction });
      const returnValue = await store.dispatch(
        actions.createInteraction(interaction)
      );
      expect(fetch.mock.calls.length).toBe(1);
      expect(fetch.mock.calls[0]).toMatchInlineSnapshot(`
        Array [
          "/api/interactions",
          Object {
            "body": "{\\"id\\":1,\\"created_at\\":\\"2021-02-25T17:31:33.826Z\\",\\"updated_at\\":\\"2021-02-25T20:09:04.999Z\\",\\"deleted_at\\":null,\\"is_active\\":true,\\"title\\":\\"Emoji Analysis\\",\\"name\\":\\"Fake Interaction\\",\\"description\\":\\"Detects the presense of an emoji character in your text\\",\\"endpoint\\":\\"ws://emoji-analysis-production.herokuapp.com\\",\\"configuration\\":{\\"bar\\":\\"2\\",\\"baz\\":\\"c\\",\\"foo\\":\\"false\\"},\\"interaction\\":{\\"id\\":1,\\"name\\":\\"ChatPrompt\\",\\"description\\":\\"It will appear as an option for scenario authors to include in chat discussions within multi-participant scenarios. It receives participant chat messages.\\",\\"created_at\\":\\"2021-02-25T15:09:05.001302-05:00\\",\\"deleted_at\\":null,\\"updated_at\\":null},\\"owner\\":{\\"id\\":999,\\"email\\":\\"super@email.com\\",\\"roles\\":[\\"participant\\",\\"super_admin\\",\\"facilitator\\",\\"researcher\\"],\\"is_super\\":true,\\"username\\":\\"superuser\\",\\"is_anonymous\\":false,\\"personalname\\":\\"Super User\\"},\\"self\\":{\\"id\\":148,\\"email\\":null,\\"roles\\":null,\\"is_super\\":false,\\"username\\":\\"ebe565050b31cbb4e7eacc39b23e2167\\",\\"lastseen_at\\":\\"2021-02-25T13:08:57.323-05:00\\",\\"is_anonymous\\":true,\\"personalname\\":\\"Emoji Analysis\\",\\"single_use_password\\":false}}",
            "headers": Object {
              "Content-Type": "application/json",
            },
            "method": "POST",
          },
        ]
      `);
      expect(returnValue).toEqual(interaction);

      await mockStore.dispatch(actions.createInteraction(interaction));
      expect(mockStore.getActions()).toMatchSnapshot();
    });

    test('Empty params', async () => {
      fetchImplementation(fetch, 200, { interaction });
      const returnValue = await store.dispatch(actions.createInteraction({}));
      expect(fetch.mock.calls.length).toBe(0);
      expect(returnValue).toEqual(null);

      await mockStore.dispatch(actions.setInteraction(1, {}));
      expect(mockStore.getActions()).toMatchSnapshot();
    });
  });
});

describe('GET_INTERACTION_ERROR', () => {
  let interaction = { ...state.interactions[0], name: 'Fake Interaction' };

  describe('getInteraction', () => {
    test('Receives an error', async () => {
      fetchImplementation(fetch, 200, { error });
      const returnValue = await store.dispatch(actions.getInteraction(1));
      expect(fetch.mock.calls.length).toBe(1);
      expect(fetch.mock.calls[0]).toMatchInlineSnapshot(`
        Array [
          "/api/interactions/1",
        ]
      `);
      expect(returnValue).toEqual(null);

      await mockStore.dispatch(actions.getInteraction(1));
      expect(mockStore.getActions()).toMatchSnapshot();
    });
  });

  describe('createInteraction', () => {
    test('Receives an error', async () => {
      fetchImplementation(fetch, 200, { error });
      const returnValue = await store.dispatch(
        actions.createInteraction(interaction)
      );
      expect(fetch.mock.calls.length).toBe(1);
      expect(fetch.mock.calls[0]).toMatchInlineSnapshot(`
        Array [
          "/api/interactions",
          Object {
            "body": "{\\"id\\":1,\\"created_at\\":\\"2021-02-25T17:31:33.826Z\\",\\"updated_at\\":\\"2021-02-25T20:09:04.999Z\\",\\"deleted_at\\":null,\\"is_active\\":true,\\"title\\":\\"Emoji Analysis\\",\\"name\\":\\"Fake Interaction\\",\\"description\\":\\"Detects the presense of an emoji character in your text\\",\\"endpoint\\":\\"ws://emoji-analysis-production.herokuapp.com\\",\\"configuration\\":{\\"bar\\":\\"2\\",\\"baz\\":\\"c\\",\\"foo\\":\\"false\\"},\\"interaction\\":{\\"id\\":1,\\"name\\":\\"ChatPrompt\\",\\"description\\":\\"It will appear as an option for scenario authors to include in chat discussions within multi-participant scenarios. It receives participant chat messages.\\",\\"created_at\\":\\"2021-02-25T15:09:05.001302-05:00\\",\\"deleted_at\\":null,\\"updated_at\\":null},\\"owner\\":{\\"id\\":999,\\"email\\":\\"super@email.com\\",\\"roles\\":[\\"participant\\",\\"super_admin\\",\\"facilitator\\",\\"researcher\\"],\\"is_super\\":true,\\"username\\":\\"superuser\\",\\"is_anonymous\\":false,\\"personalname\\":\\"Super User\\"},\\"self\\":{\\"id\\":148,\\"email\\":null,\\"roles\\":null,\\"is_super\\":false,\\"username\\":\\"ebe565050b31cbb4e7eacc39b23e2167\\",\\"lastseen_at\\":\\"2021-02-25T13:08:57.323-05:00\\",\\"is_anonymous\\":true,\\"personalname\\":\\"Emoji Analysis\\",\\"single_use_password\\":false}}",
            "headers": Object {
              "Content-Type": "application/json",
            },
            "method": "POST",
          },
        ]
      `);
      expect(returnValue).toEqual(null);

      await mockStore.dispatch(actions.createInteraction(interaction));
      expect(mockStore.getActions()).toMatchSnapshot();
    });
  });
});

describe('SET_INTERACTION_SUCCESS', () => {
  let interaction = { ...state.interactions[0], name: 'Fake Interaction' };

  describe('setInteraction', () => {
    test('Empty params', async () => {
      fetchImplementation(fetch, 200, { interaction });
      const returnValue = await store.dispatch(actions.setInteraction(1, {}));
      expect(fetch.mock.calls.length).toBe(0);
      expect(returnValue).toEqual(null);

      await mockStore.dispatch(actions.setInteraction(1, {}));
      expect(mockStore.getActions()).toMatchSnapshot();
    });

    test('Receives a interaction', async () => {
      fetchImplementation(fetch, 200, { interaction });
      const returnValue = await store.dispatch(
        actions.setInteraction(1, interaction)
      );
      expect(fetch.mock.calls.length).toBe(1);
      expect(fetch.mock.calls[0]).toMatchInlineSnapshot(`
        Array [
          "/api/interactions/1",
          Object {
            "body": "{\\"id\\":1,\\"created_at\\":\\"2021-02-25T17:31:33.826Z\\",\\"updated_at\\":\\"2021-02-25T20:09:04.999Z\\",\\"deleted_at\\":null,\\"is_active\\":true,\\"title\\":\\"Emoji Analysis\\",\\"name\\":\\"Fake Interaction\\",\\"description\\":\\"Detects the presense of an emoji character in your text\\",\\"endpoint\\":\\"ws://emoji-analysis-production.herokuapp.com\\",\\"configuration\\":{\\"bar\\":\\"2\\",\\"baz\\":\\"c\\",\\"foo\\":\\"false\\"},\\"interaction\\":{\\"id\\":1,\\"name\\":\\"ChatPrompt\\",\\"description\\":\\"It will appear as an option for scenario authors to include in chat discussions within multi-participant scenarios. It receives participant chat messages.\\",\\"created_at\\":\\"2021-02-25T15:09:05.001302-05:00\\",\\"deleted_at\\":null,\\"updated_at\\":null},\\"owner\\":{\\"id\\":999,\\"email\\":\\"super@email.com\\",\\"roles\\":[\\"participant\\",\\"super_admin\\",\\"facilitator\\",\\"researcher\\"],\\"is_super\\":true,\\"username\\":\\"superuser\\",\\"is_anonymous\\":false,\\"personalname\\":\\"Super User\\"},\\"self\\":{\\"id\\":148,\\"email\\":null,\\"roles\\":null,\\"is_super\\":false,\\"username\\":\\"ebe565050b31cbb4e7eacc39b23e2167\\",\\"lastseen_at\\":\\"2021-02-25T13:08:57.323-05:00\\",\\"is_anonymous\\":true,\\"personalname\\":\\"Emoji Analysis\\",\\"single_use_password\\":false}}",
            "headers": Object {
              "Content-Type": "application/json",
            },
            "method": "PUT",
          },
        ]
      `);
      expect(returnValue).toEqual(interaction);

      await mockStore.dispatch(actions.setInteraction(1, interaction));
      expect(mockStore.getActions()).toMatchSnapshot();
    });

    test('Receives a interaction, updates scenario', async () => {
      store = createMockConnectedStore({
        ...original,
        scenario: {
          ...original.scenario,
          interactions: [...original.interactions]
        }
      });

      fetchImplementation(fetch, 200, { interaction });
      const returnValue = await store.dispatch(
        actions.setInteraction(1, interaction)
      );
      expect(fetch.mock.calls.length).toBe(1);
      expect(fetch.mock.calls[0]).toMatchInlineSnapshot(`
        Array [
          "/api/interactions/1",
          Object {
            "body": "{\\"id\\":1,\\"created_at\\":\\"2021-02-25T17:31:33.826Z\\",\\"updated_at\\":\\"2021-02-25T20:09:04.999Z\\",\\"deleted_at\\":null,\\"is_active\\":true,\\"title\\":\\"Emoji Analysis\\",\\"name\\":\\"Fake Interaction\\",\\"description\\":\\"Detects the presense of an emoji character in your text\\",\\"endpoint\\":\\"ws://emoji-analysis-production.herokuapp.com\\",\\"configuration\\":{\\"bar\\":\\"2\\",\\"baz\\":\\"c\\",\\"foo\\":\\"false\\"},\\"interaction\\":{\\"id\\":1,\\"name\\":\\"ChatPrompt\\",\\"description\\":\\"It will appear as an option for scenario authors to include in chat discussions within multi-participant scenarios. It receives participant chat messages.\\",\\"created_at\\":\\"2021-02-25T15:09:05.001302-05:00\\",\\"deleted_at\\":null,\\"updated_at\\":null},\\"owner\\":{\\"id\\":999,\\"email\\":\\"super@email.com\\",\\"roles\\":[\\"participant\\",\\"super_admin\\",\\"facilitator\\",\\"researcher\\"],\\"is_super\\":true,\\"username\\":\\"superuser\\",\\"is_anonymous\\":false,\\"personalname\\":\\"Super User\\"},\\"self\\":{\\"id\\":148,\\"email\\":null,\\"roles\\":null,\\"is_super\\":false,\\"username\\":\\"ebe565050b31cbb4e7eacc39b23e2167\\",\\"lastseen_at\\":\\"2021-02-25T13:08:57.323-05:00\\",\\"is_anonymous\\":true,\\"personalname\\":\\"Emoji Analysis\\",\\"single_use_password\\":false}}",
            "headers": Object {
              "Content-Type": "application/json",
            },
            "method": "PUT",
          },
        ]
      `);
      expect(returnValue).toEqual(interaction);

      await mockStore.dispatch(actions.setInteraction(1, interaction));
      expect(mockStore.getActions()).toMatchSnapshot();
    });
  });
});

describe('SET_INTERACTION_ERROR', () => {
  let interaction = { ...state.interactions[0], name: 'Fake Interaction' };

  describe('setInteraction', () => {
    test('Receives an error', async () => {
      fetchImplementation(fetch, 200, { error });
      const returnValue = await store.dispatch(
        actions.setInteraction(1, interaction)
      );
      expect(fetch.mock.calls.length).toBe(1);
      expect(fetch.mock.calls[0]).toMatchInlineSnapshot(`
        Array [
          "/api/interactions/1",
          Object {
            "body": "{\\"id\\":1,\\"created_at\\":\\"2021-02-25T17:31:33.826Z\\",\\"updated_at\\":\\"2021-02-25T20:09:04.999Z\\",\\"deleted_at\\":null,\\"is_active\\":true,\\"title\\":\\"Emoji Analysis\\",\\"name\\":\\"Fake Interaction\\",\\"description\\":\\"Detects the presense of an emoji character in your text\\",\\"endpoint\\":\\"ws://emoji-analysis-production.herokuapp.com\\",\\"configuration\\":{\\"bar\\":\\"2\\",\\"baz\\":\\"c\\",\\"foo\\":\\"false\\"},\\"interaction\\":{\\"id\\":1,\\"name\\":\\"ChatPrompt\\",\\"description\\":\\"It will appear as an option for scenario authors to include in chat discussions within multi-participant scenarios. It receives participant chat messages.\\",\\"created_at\\":\\"2021-02-25T15:09:05.001302-05:00\\",\\"deleted_at\\":null,\\"updated_at\\":null},\\"owner\\":{\\"id\\":999,\\"email\\":\\"super@email.com\\",\\"roles\\":[\\"participant\\",\\"super_admin\\",\\"facilitator\\",\\"researcher\\"],\\"is_super\\":true,\\"username\\":\\"superuser\\",\\"is_anonymous\\":false,\\"personalname\\":\\"Super User\\"},\\"self\\":{\\"id\\":148,\\"email\\":null,\\"roles\\":null,\\"is_super\\":false,\\"username\\":\\"ebe565050b31cbb4e7eacc39b23e2167\\",\\"lastseen_at\\":\\"2021-02-25T13:08:57.323-05:00\\",\\"is_anonymous\\":true,\\"personalname\\":\\"Emoji Analysis\\",\\"single_use_password\\":false}}",
            "headers": Object {
              "Content-Type": "application/json",
            },
            "method": "PUT",
          },
        ]
      `);
      expect(returnValue).toEqual(null);

      await mockStore.dispatch(actions.setInteraction(1, interaction));
      expect(mockStore.getActions()).toMatchSnapshot();
    });
  });
});

describe('GET_INTERACTIONS_SUCCESS', () => {
  describe('getInteractions', () => {
    let interactions = [...state.interactions];

    test('Receives interactions', async () => {
      fetchImplementation(fetch, 200, { interactions });
      const returnValue = await store.dispatch(actions.getInteractions());
      expect(fetch.mock.calls.length).toBe(1);
      expect(fetch.mock.calls[0]).toMatchInlineSnapshot(`
        Array [
          "/api/interactions",
        ]
      `);
      expect(returnValue).toEqual(interactions);

      await mockStore.dispatch(actions.getInteractions());
      expect(mockStore.getActions()).toMatchSnapshot();
    });

    test('Receives undefined', async () => {
      fetchImplementation(fetch, 200, { interactions: undefined });
      const returnValue = await store.dispatch(actions.getInteractions());
      expect(fetch.mock.calls.length).toBe(1);
      expect(fetch.mock.calls[0]).toMatchInlineSnapshot(`
        Array [
          "/api/interactions",
        ]
      `);
      expect(returnValue).toMatchInlineSnapshot(`Array []`);

      await mockStore.dispatch(actions.getInteractions());
      expect(mockStore.getActions()).toMatchSnapshot();
    });
  });
});

describe('GET_INTERACTIONS_ERROR', () => {
  describe('getInteractions', () => {
    test('Receives error', async () => {
      fetchImplementation(fetch, 200, { error });
      const returnValue = await store.dispatch(actions.getInteractions());
      expect(fetch.mock.calls.length).toBe(1);
      expect(returnValue).toEqual(null);
      await mockStore.dispatch(actions.getInteractions());
      expect(mockStore.getActions()).toMatchSnapshot();
    });
  });
});
