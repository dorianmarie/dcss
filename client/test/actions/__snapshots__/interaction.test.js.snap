// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GET_INTERACTION_ERROR createInteraction Receives an error 2`] = `
Array [
  Object {
    "error": Object {
      "error": [Error: something unexpected happened on the server],
    },
    "type": "GET_INTERACTION_ERROR",
  },
]
`;

exports[`GET_INTERACTION_ERROR getInteraction Receives an error 2`] = `
Array [
  Object {
    "error": Object {
      "error": [Error: something unexpected happened on the server],
    },
    "type": "GET_INTERACTION_ERROR",
  },
]
`;

exports[`GET_INTERACTION_SUCCESS createInteraction Empty params 1`] = `Array []`;

exports[`GET_INTERACTIONS_ERROR getInteractions Receives error 1`] = `
Array [
  Object {
    "error": Object {
      "error": [Error: something unexpected happened on the server],
    },
    "type": "GET_INTERACTIONS_ERROR",
  },
]
`;

exports[`GET_INTERACTIONS_SUCCESS getInteractions Receives interactions 2`] = `
Array [
  Object {
    "interactions": Array [
      Object {
        "created_at": "2021-02-25T17:31:33.826Z",
        "deleted_at": null,
        "description": "It will appear as an option for scenario authors to include in chat discussions within multi-participant scenarios. It receives participant chat messages.",
        "id": 1,
        "name": "ChatPrompt",
        "owner": Object {
          "email": "super@email.com",
          "id": 999,
          "is_anonymous": false,
          "is_super": true,
          "personalname": "Super User",
          "roles": Array [
            "participant",
            "super_admin",
            "facilitator",
            "researcher",
          ],
          "username": "superuser",
        },
        "updated_at": "2021-02-27T20:38:53.774Z",
      },
      Object {
        "created_at": "2021-02-25T17:31:33.826Z",
        "deleted_at": null,
        "description": "It will appear as an option for scenario authors to use in conditional content components within scenarios. It receives participant Audio Prompt Responses.",
        "id": 2,
        "name": "AudioPrompt",
        "owner": Object {
          "email": "super@email.com",
          "id": 999,
          "is_anonymous": false,
          "is_super": true,
          "personalname": "Super User",
          "roles": Array [
            "participant",
            "super_admin",
            "facilitator",
            "researcher",
          ],
          "username": "superuser",
        },
        "updated_at": "2021-02-27T20:38:53.774Z",
      },
      Object {
        "created_at": "2021-02-25T17:31:33.826Z",
        "deleted_at": null,
        "description": "It will appear as an option for scenario authors to use in conditional content components within scenarios. It receives participant Text Prompt Responses.",
        "id": 3,
        "name": "TextPrompt",
        "owner": Object {
          "email": "super@email.com",
          "id": 999,
          "is_anonymous": false,
          "is_super": true,
          "personalname": "Super User",
          "roles": Array [
            "participant",
            "super_admin",
            "facilitator",
            "researcher",
          ],
          "username": "superuser",
        },
        "updated_at": "2021-02-27T20:38:53.774Z",
      },
    ],
    "type": "GET_INTERACTIONS_SUCCESS",
  },
]
`;

exports[`GET_INTERACTIONS_SUCCESS getInteractions Receives undefined 3`] = `
Array [
  Object {
    "interactions": undefined,
    "type": "GET_INTERACTIONS_SUCCESS",
  },
]
`;

exports[`SET_INTERACTION_ERROR setInteraction Receives an error 2`] = `
Array [
  Object {
    "error": Object {
      "error": [Error: something unexpected happened on the server],
    },
    "type": "SET_INTERACTION_ERROR",
  },
]
`;

exports[`SET_INTERACTION_SUCCESS setInteraction Empty params 1`] = `Array []`;
